### DISCUSSION TITLE
Перенос плейбэка на любое устройство

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#126] [Помощь с алгоритмом] [answered]
Дошли руки и до этой функции.
Опять же после появления возможности использовать Таскер, пытаюсь реализовать давнюю идею, с ее применением.

Часто при включении ГУ в машине (в нем полноценный клиент Spotify на Андроиде) приходится делать принудительную синхронизацию с телефоном, для того чтобы продолжить воспроизведение на текущем месте. Каждый раз реакция девайсов довольно разная. То один из них перехватывает воспроизведение и управление другим с помощью Spotify connect, то другой. В общем не удобно все это.
Пытаюсь с помощью Гуфи и Таскера автоматизировать процесс. На данном этапе отлично работает принудительное включение конкретного плейлиста ( плейлиста дня Smartify) на ГУ, при подключении телефона к нему по ВТ.
```
function Auto() {
  let playlistId = '74uSA9FulwvWqVSi7mxWyQ';

  Player.resume({
  deviceId: '8471d89fd60a8eab0289c3a3339e946a474e15cc',  context_uri: `spotify:playlist:${playlistId}`,
});
}
```
Но это частный случай. Заинтересовала возможность включения любого плейлиста ,  игравшего последним на любом из устройств. Можно ли используя эту функцию на конкретном устройстве стартовать любой, а не конкретный плейлист? Тест ее запуска без указания ID плейлиста дает ошибку.

**Comment by Chimildic** [comment, marked as answer]
Звучит конечно как костыль для костылей) Точно не решается каким-нибудь переключателем в настройках или последовательностью включения?

Данные о последнем играющем треке или плейлисте живут на паузе очень мало. Имею ввиду `getPlayback`. 

Допустим трек играет в телефоне, тогда по триггеру блютуза плейбэк можно перекинуть в машину. [Этой функции](https://developer.spotify.com/documentation/web-api/reference/#endpoint-transfer-a-users-playback) нет в библиотеки, но она короткая. По идеи тоже самое делает сам Spotify, когда ловит новое устройство.
```js
SpotifyRequest.put('https://api.spotify.com/v1/me/player', {
  device_ids: ['8471d89fd60a8eab0289c3a3339e946a474e15cc'],
  play: true,
});
```

`Player.resume` сработает без ошибки, когда сам плейбэк еще _жив_. Остается предложить экзотику в виде кэша, который бы хранил последнее состояние плейбэка для восстановления. Но возникает вопрос его актуальности. Как часто опрашивать и каким способом? Можно запросить плейбэк у Spotify. Если ответ пустой, то взять из кэша. 

Конечный вариант в [комментарии](https://github.com/Chimildic/goofy/discussions/126#discussioncomment-1316319).

**Comment by Bedmit** [comment]
Вернулся к экспериментам )
`SpotifyRequest` отлично работает. Даже сделал себе серию функций переключающих воспроизведение на любой свой девайс. Оказалось даже через веб форму быстрее и удобней 
![Screenshot_20210913-163324__01](https://user-images.githubusercontent.com/76230194/133093759-7b7865ae-90cf-49ca-95d6-ae873aa8d55d.jpg)
А вообще планирую это добро на управление голосом заточить.
Это вынужденный, но как оказалось единственный рабочий костыль. 

Просто у Spotify connect довольно странный алгоритм работы. Он начинает работать не при открытии приложения, а при запуске трека. Если открыть давно игравшее приложение на каком либо устройстве, оно само не соединится с тем, что играло последним, даже если то еще висит в фоне. Включив воспроизведение, вместо продолжения того что слушал ранее, получаем то, что слушал последний раз на этом устройстве. 
Если нужно продолжить воспроизведение, что требуется в 99% случаев, нужно сперва запустить клиент последнего игравшего устройства, включить клиент на котором будет продолжено воспроизведение, подождать пока все синхронизируется и переключиться на нужное устройство.
Т.е. простое принудительное включение воспроизведения на устройстве (тем же Таскером) здесь не подходит. Оно включит старый плейлист, тот что остался в кэше устройства. А после этого еще и синхронизируется с предыдущим девайсом и включив этот плейлист и там.

А `SpotifyRequest` как раз и делает то что надо - запускает то, что находится в плейбэке. Меня это более чем устраивает без всяких кэшей. А если плейбэка уже нет, то тогда пусть включается мой Smartify '74uSA9FulwvWqVSi7mxWyQ' 
В качестве жирной точки в вопросе, прошу помочь правильно написать это условие.

Пока решил это так
```
let playback = Player.getPlayback();
  if (playback.isEmpty()) {
  SpotifyRequest.put('https://api.spotify.com/v1/me/player', {
  device_ids: ['8471d89fd60a8eab0289c3a3339e946a474e15cc'],
  play: true,
  });
  return;
  } else {
  let playlistId = '74uSA9FulwvWqVSi7mxWyQ';
    Player.resume({
      deviceId: '8471d89fd60a8eab0289c3a3339e946a474e15cc',
      context_uri: `spotify:playlist:${playlistId}`,
   });
}
```

**Reply by Chimildic** [reply]
Добавил [transferPlayback](https://chimildic.github.io/goofy/#/func?id=transferplayback) в библиотеку. 

> правильно написать это условие

```js
function Auto() {
  const SMARTIFY_ID = '74uSA9FulwvWqVSi7mxWyQ';
  const AUTO_ID = '8471d89fd60a8eab0289c3a3339e946a474e15cc';

  let playback = Player.getPlayback();
  if (playback.isEmpty()) {
    Player.resume({ deviceId: AUTO_ID, context_uri: `spotify:playlist:${SMARTIFY_ID}` });
  } else {
    Player.transferPlayback(AUTO_ID, true);
  }
}
```

Нам пока неизвестно меняет ли Spotify у устройств их _id_. Как выяснилось в какой-то ветки, может изменить у трека или исполнителя. Правда довольно редко, судя по количество. Я к тому, что возможно понадобится динамически узнавать _id_ машины. Как пример ниже. При этом на устройстве должен быть включен Spotify.
```js
let device = Player.getAvailableDevices().find(d => d.type == 'Smartphone');
// device.id
```
Но я не знаю какой тип у машины. Возможно `Auto`. Более точно узнать можно при дебаге или выводе в консоль вызова `Player.getAvailableDevices()`.

**Reply by Bedmit** [reply]
> Но я не знаю какой тип у машины

На балалайках с Андроид, планшетная - Tablet

**Reply by Bedmit** [reply]
> ```js
> Player.transferPlayback(AUTO_ID, true);
> ```

При false я так понял происходит переключение на девайс, но в режиме паузы?

**Reply by Chimildic** [reply]
> но в режиме паузы

Spotify пишет, что при _false_ сохранит состояние с предыдущего устройства. Надо на практике проверить.

**Reply by Bedmit** [reply]
Да, проверил, так и есть. Довольно удобно. 

Протестировал новые функции. Опять доволен как слон. Спасибо.
Боюсь это не последняя безумная идея ))

**Reply by Bedmit** [reply]
Опять заблудился в трех соснах.
Теперь нужна обратная задача. Нужно запустить плейлист Smartify на текущем устройстве. Если Smartify есть в плэйбеке, то просто продолжить с того же места. Если это не Smartify, или плэйбек пустой, то запустить Smartify.
```js
function SmartifyStart() {
  const SMARTIFY_ID = '74uSA9FulwvWqVSi7mxWyQ';

  let playback = Player.getPlayback();
  if (playback.isEmpty()
    || playback.context.uri.includes(SMARTIFY_ID)) {
    Player.resume();
  } else {
    Player.resume({ context_uri: `spotify:playlist:${SMARTIFY_ID}` });
  }
}
```
Вроде все просто, но при пустом плэйбеке выдает ошибку 404

**Reply by Chimildic** [reply]
Многовато становится функций, их там никак не объединить? Может описать полный сценарий использования, чтобы лишь условия делать, а не функции.

Не правильно воспринимается условие. Функция `isEmpty` даст `true` когда плейбэк пуст. При `true` срабатывает первая ветка `if`, а не ветка `else`. Восклицательный знак в условии `!` это инверсия булева значения. Кстати, если играют любимые треки, то в контексте не будет `uri` и код оборвется ошибкой
```js
if (playback.isEmpty() || !playback.context || !playback.context.uri.includes(SMARTIFY_ID)){
  // плейбэк пустой или нет контекста или не smartify
  Player.resume({ context_uri: `spotify:playlist:${SMARTIFY_ID}` });
} else {
  // плейбэк не пустой, играет smartify
}
```
Наверное еще нужны условия во второй блок. Потому что ловлю ошибку 403, когда smartify играет на том же устройстве.

**Reply by Bedmit** [reply]
> их там никак не объединить?

Не знаю. Мне это все для запуска Таскером нужно. Поэтому все делаю отдельными функциями. Если Таскер может еще выполнять задачу и по изменению условий, то почему бы и нет.

Но там разные задачи. Если первый случай - переключение на устройство и запуск на нем текущего плэйбека, или Smartify при его отсутствии. 
А второй - запуск на текущем устройстве Smartify, если он ни в плэйбеке. Если там Smartify, то продолжение прослушивания. Если в этом случае просто запускать Smartify, то он не продолжает прослушивание, а запускает его сначала.

 

> ловлю ошибку 403, когда smartify играет на том же устройстве

Точно. Только сейчас обратил внимание. Но кроме сообщения это ни как не напрягает.



> Наверное еще нужны условия во второй блок

Тоже об этом подумал

**Reply by Bedmit** [reply]
А кстати. Можно как-то принудительно выгрузить плэйбек? Для тестирования скрипта.

**Reply by Chimildic** [reply]
> Можно как-то принудительно выгрузить плэйбек?

Попробовал закрыть приложение везде, плейбэк пустой стал сразу
